javascript/photo.js

```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
+        var order = localStorage.getItem("order");
+        var cachedImages = localStorage.getItem("cachedImages");
+        return count != null && Number(count) >= 0 && order != null && JSON.parse(order).length == this.BACKUP_IMAGES_COUNT && cachedImages != null && JSON.parse(cachedImages).length <= cacheSize;
+    },
+
     /**
      * Setup LocalStorage before displaying first image.
      */
     setup: function() {
         var order = this.permutation(this.BACKUP_IMAGES_COUNT);
         localStorage.setItem("count", 0);
-        localStorage.setItem("cacheIndex", 0);
         localStorage.setItem("order", JSON.stringify(order));
         localStorage.setItem("cachedImages", JSON.stringify([]));
     },
 
     /**
      * Displays the image.
-     * TODO: Update function name & doc. This function does more than just display
+     * TODO: Update function name & doc. This function does more than just
+     * display
      */
     display: function(numberOfImages, cacheSize) {
-        var count = localStorage.getItem("count");
-        if (count === null) {
+        if (!this.isValidLocalStorageState(cacheSize)) {
             this.setup();
         }
-        count = Number(count);
-        var cacheIndex = Number(localStorage.getItem("cacheIndex"));
+        // From this point onwards, count, order and cachedImages are in a valid state (to know what a valid state is, refer to the IsValidLocalStorage function).
+        var count = Number(localStorage.getItem("count"));
         var order = JSON.parse(localStorage.getItem("order"));
-        if (count < this.STARTUP_IMAGE_COUNT) {
-            this.displayImage(this.offlineImages[order[count]]);
+        var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
+        // We display backup images in the following cases.
+        // * count < STARTUP_IMAGE_COUNT: We are yet to finish the quota of offline images before we can use images from the cache.
+        // * count >= STARTUP_IMAGE_COUNT but cachedImages.length == 0: Can happen when HTTP requests to r/EarthPorn fail. We fall back to display the backup images.
+        if (count < this.STARTUP_IMAGE_COUNT || cachedImages.length == 0) {
+            var order_index = count % this.BACKUP_IMAGES_COUNT;
+            this.displayImage(this.offlineImages[order[order_index]]);
         } else {
-            // TODO: Use variable cacheReadPos & cacheWritePos to prevent errors when fetch fails
-            var cachedImageNumber = cacheIndex - this.STARTUP_IMAGE_COUNT;
-            if (cachedImageNumber < 0) {
-                cachedImageNumber += cacheSize;
-            }
-            var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
-            this.displayImage(cachedImages[cachedImageNumber]);
+            this.displayImage(cachedImages[0]);
+            // Once we consume an image from the cache, remove it from the cache.
+            cachedImages.splice(0, 1);
```

Comment:
When will the cache size be > this.STARTUP_IMAGE_COUNT? cacheSize variable seems unnecessary.

Reply:
STARTUP_IMAGE_COUNT and cacheSize shouldn't have any dependency on one another.  For example, STARTUP_IMAGE_COUNT could be equal to BACKUP_IMAGES_COUNT and that shouldn't break how we use the cache.